Supervisor Framework

jsiwek opened the pull request:
This is the initial infrastructure for the new process Supervisor mode of Zeek (#582)

* Documentation in `topic/jsiwek/supervisor` of the `zeek-docs` repo needs to be merged with this: see https://github.com/zeek/zeek-docs/pull/20
* Doesn't support offline pcap reading yet, can plan to add it for 3.2.0
* Assuming CI passes and there's only Good Feedback, this could be merged right away, but I expect conflicts with #681, so can either wait for that and I'll push resolutions here or else ask me to help resolve those if needed when it's time to merge that other branch

rsmmr commented on src/main.cc:
```diff
@@ -94,10 +99,10 @@ file_analysis::Manager* file_mgr = 0;
 iosource::Manager* iosource_mgr = 0;
 bro_broker::Manager* broker_mgr = 0;
+zeek::Supervisor* zeek::supervisor = 0;
+std::optional<zeek::Supervisor::SupervisedNode> zeek::supervised_node;
```
Move this into the Supervisor as a static function returning the node object?

jsiwek replied:
Yeah, sounds good.

rsmmr commented on src/main.cc:
```diff
@@ -203,18 +208,533 @@ void usage(int code = 1)
+		/* interfaces = og.interfaces; */
+		/* pcap_files = og.pcap_files; */
+
+		pcap_output_file = og.pcap_output_file;
```
I assume this end up in the processes own working directory so that files don't conflict? (Same for other options creating files)

jsiwek replied:
Right, nothing special here -- if one uses a distinct working directory for supervised nodes then they avoid any file clobbering/conflicts.  If they use the same shared working dir, they risk potential conflicts.

rsmmr commented on src/main.cc:
```diff
@@ -203,18 +208,533 @@ void usage(int code = 1)
+
+		pcap_output_file = og.pcap_output_file;
+		random_seed_input_file = og.random_seed_input_file;
+		random_seed_output_file = og.random_seed_output_file;
```
Probably not much use having this inherited by  the child processes?

jsiwek replied:
It's maybe less likely to see legit use, but don't see why not?  Also, the more options we stop from being inherited, the more complex documenting/explaining that inheritance becomes and here just seems easier to let it be inherited.

rsmmr commented on src/main.cc:
```diff
@@ -203,18 +208,533 @@ void usage(int code = 1)
+		{
+		if ( chdir(config.directory->data()) )
+			{
+			fprintf(stderr, "node '%s' failed to chdir to %s: %s\n",
```
could these error messages go back the supervisor to be logged? Not quite sure where in the set up process we are here.

jsiwek replied:
That's on the supervised (Leaf) node.  There's not currently any IPC between Leaf and Stem (or Leaf and Supervisor), just between Stem and Supervisor processes.  Can try adding if you want, but also not sure where to draw the line on what messages to report back: (1) just supervisor options that results in errors at init-time (2) any error during init (3) any error during run-time, etc.

rsmmr replied:
ok for now, let's see if this comes up as an issue

rsmmr commented on src/main.cc:
```diff
@@ -203,18 +208,533 @@ void usage(int code = 1)
+		}
+};
+
+static void init_supervised_node(zeek_options* options)
```
Move this out of `main.cc` into the supervisor code?

jsiwek replied:
Sure (mostly only there due to `zeek_options` dependency, but since that's also getting factored...)

rsmmr commented on src/main.cc:
```diff
@@ -203,18 +208,533 @@ void usage(int code = 1)
 	exit(code);
 	}
 
+struct zeek_options {
```
with the option stuff being factored out, move out of main.cc into a separate file?

jsiwek replied:
I wasn't too motivated to do that since it's still an implementation detail of the arg parsing still not needing use/exposure anywhere in public API, but sure, I'll factor into `zeek::Options` in an `Options.h` and we can review proper way to expose and/or filter which headers get installed later.

rsmmr commented on src/main.cc:
```diff
@@ -407,327 +928,183 @@ static void bro_new_handler()
+	}
 
-	string zeekygen_config;
+static std::string get_exe_path(const std::string& invocation)
```
Could this just use `realpath`?

jsiwek replied:
Don't think so -- the more important thing being done here is the `PATH` lookup.  e.g. find out to what an invocation via just `zeek` refers.  Nice to preserve the absolute path in case it needs to `exec()` a new Stem since it's otherwise just another way to pick up a different binary (if a different one got inserted at an early point in `PATH` before the `exec()` happens)

rsmmr commented on src/main.cc:
```diff
@@ -407,327 +928,183 @@ static void bro_new_handler()
-			fprintf(stdout, "%s version %s\n", prog, zeek_version());
-			exit(0);
-			break;
+	pid_t stem_pid = 0;
```
Add some comments here what's going on with the stem process & env variable, not clear without also reading the supervisor code.

Also, move more functionality into the Supervisor?

jsiwek replied:
Sure, agree it's better organizing all that behind a single `Supervisor::CreateStem()` call.

jsiwek commented on scripts/base/frameworks/cluster/nodes/logger.zeek:
```diff
@@ -23,4 +23,7 @@ redef Log::default_rotation_interval = 1 hrs;
 
 ## Use the cluster's archive logging script.
+
+@if ( ! Supervisor::is_supervised() )
```
Since most likely no one will have the `archive-log` script in their `PATH`, file rotation will emit errors that it can't be found.  I'm thinking that `redef` doesn't quite belong there in the first place -- if it's a contrivance for ZeekControl to have set up `PATH`, it could have instead generated its own script with that `redef` (could even generate it with absolute path if it wanted).

jsiwek commented on src/zeek-affinity.h:
```diff
@@ -0,0 +1,15 @@
+// See the file "COPYING" in the main distribution directory for copyright.
+
```
I'll add a comment in the source file to explain why it's independence is warranted:

> // This is all in its own source file primarily because the Linux
// implementation uses the _GNU_SOURCE feature test macro which must be
// defined before including any header file and lumping this together with
// other util functions makes that requirement less apparent and less
// self-contained.

Doesn't necessarily need an independent header file, but seemed the most logical way to organize and also think this function is fine to expose publicly, e.g. for external plugin use if they want.  Let me know if you had other ideas.

jsiwek commented on src/main.cc:
```diff
@@ -94,10 +99,10 @@ file_analysis::Manager* file_mgr = 0;
 zeekygen::Manager* zeekygen_mgr = 0;
 iosource::Manager* iosource_mgr = 0;
 bro_broker::Manager* broker_mgr = 0;
+zeek::Supervisor* zeek::supervisor = 0;
```
Sure, will rename.

jsiwek commented:
@rsmmr thanks for the feedback, pushed new commits to address most code comments.

> * I'm wondering if it would make sense to split the script-side API into to parts: functions to call from inside the supervisor, and functions to control the supervisor from other processes. Right now the distinction in only made in comments ("It's an error to ..."). And/or: at least factor our the "remote control API" through events as that feels a bit unnatural for a framework's API. For the framework, it would be more natural to have functions to call instead of the `*_request` events(). We could also offer functions on top of the events, and use events only for external control.

Have any concrete changes in mind?

The request/response events seem like the lowest-level building block that most closely models the async behavior we need for this functionality.

* If one wanted a convenience function that generates the request, they can still make that themselves.  The return value would be kind of meaningless, though, unless we actually made it synchronous/blocking.
* I would probably find it more awkward to have the `request` hidden behind a function call because then if I want the handle the `response`, it's not symmetric anymore -- call this function to make a request, but then use an event to handle the response.
* I think we'd just end up using events to model the internal communication anyway.
* Don't think we want to hide the details of the request event since someone can actually choose to handle that at different priorities to whatever effect they need -- e.g. run some logic before/after creating a new supervised process.

Could split out the events related to remote control into a separate file if that helps, but personally don't think it's worthwhile -- nice to have everything you could want to know in one place.

> * Shall we add `zeek_{init, done}_supervisor()` events? That would avoid the if-switches in `zeek_init/done`.

Not too opposed to that, but think I like it better if the supervisor's event structure doesn't differ too much from a normal Zeek process.  Having additional supervisor-specific events then begs a couple more questions (for me or also regular users): confusion whether the supervisor runs both events or just the supervisor-specific one, then also ambiguous (without reading docs) what order they come in.  Is `zeek_init()` before or after `zeek_init_supervisor()` ?

> * I'd move all the C++-side supervisor-specific code into a `supervisor/` subdirectory to avoid further cluttering the top-level `src/` directory. 

I didn't see an easy way to do that "properly" because of the way subdirectory BIFs are handled differently than top-level BIFs: `init_builtin_funcs_subdirs()` is called after `yyparse()`, but it's too late at that point since we need/want to be able to call supervisor BIFs inside `@if` directives.

We'd probably want to generally fix that BIF initialization order/dependency mess, but kind of prefer to treat that as out of scope here.  Or I could try doing some one-off kludge just to get `supervisor.bif` inside `supervisor/` but treated the same as other top-level `.bif` files, but moving 3 source files into 1 directory isn't exactly a big enough de-cluttering win to warrant weird workarounds IMO.

> Probably also use a supervisor namespace for sub pieces?

Don't think it's worth it since there's only a few sub-pieces, they're really not names I'd expect to collide with something else, and/or adding `supervisor` namespace doesn't clarify anything.  Or let me know which names/pieces you were thinking.

> * There's a quite a bit of new code in `main.cc`. Can we factor more of that out of that file? The options could go now go into their own file, and some of the early-stage supervisor logic could move into static methods of the Supervisor I believe.

Sure, more of that is factored out now.

> * I haven't looked at the implementation enough to answer this one: Does the supervisor code depend on what node types are available? It would be nice if the the cluster framework could change node types without the supervisor caring about that.

You might have to elaborate what you mean, but the public/user-facing Supervisor API is completely independent of the public/user-facing Cluster Framework API.  If there's a change in the Cluster Framework, that doesn't necessarily effect user-code written against Supervisor Framework if we can internally handle that change.  The interaction between Supervisor and Cluster Framework is all internal details at the moment: it populates `Cluster::nodes` and sets `CLUSTER_NODE` environment variable according to what the Cluster Framework needs, but a user that wants to use the Supervisor Framework doesn't have to know anything about the Cluster Framework.

> * This might be out of scope for now, but still: `-j <N>` seems relevant not only in pcap mode, but also live: why not generally control the numbers of worker with that? For now, though, it seems the `<N>` isn't used anywhere, so I'd remove the argument from the command line usage/processing.

That's from very early on and just forgot to remove the `--help` dialog about that, but yeah, plan is that `-j <W>` or even `-j <W,P,L>` could be used to instantiate the requested number of Worker, Proxy, or Logger nodes in some sane "default" cluster configuration.  It likely does generalize to either pcap or live modes and if there's demand for that convenience functionality to be available up-front, don't think it's too far away to get that working, but still slightly prefer to wait and see if the underlying building blocks are stable.

rsmmr commented:
Jon and I synced up offline, recording conclusions in comments.

rsmmr commented:
> > * I'm wondering if it would make sense to split the script-side API into to parts: functions to call from inside the supervisor, and functions to control the supervisor from other processes. 

We decided to stay with events but factor out the control-related API into its own namespace.

> > * Shall we add `zeek_{init, done}_supervisor()` events? That would avoid the if-switches in `zeek_init/done`.

Will leave as is, because: (1) the need isn't strong, (2) these could be more confusing than helpful; (3) users will eventually rarely see this anyways; and (4)  we can always add it later easily without breaking anything if it turns out it would be helpful to provide the additional semantics.

> * I'd move all the C++-side supervisor-specific code into a `supervisor/` subdirectory to avoid further cluttering the top-level `src/` directory.

We'll do that even if it means adding a hack to get the bifs working early on. Seems worth to avoid further cluttering the top-level directory.

> > * This might be out of scope for now, but still: `-j <N>` seems relevant not only in pcap mode, but also live: why not generally control the numbers of worker with that? For now, though, it seems the `<N>` isn't used anywhere, so I'd remove the argument from the command line usage/processing.

We decided to not implement the <N> for now, as the right semantics aren't clear/obvious yet.

rsmmr commented:
Thanks I'll take another look

rsmmr commented:
Finished the 2nd part of my review re/ the implementation. This is impressive! You have all the  cases covered that I can think of for the various situations the process tree might get into. In particular it's great to see the care when restarting or killing nodes, potentially multiple times and with delays.

Just a few further notes: 

1. When starting the stem process: given that we have use exec + ZEEK_STEM anyways when reviving the stem, we could switch to always using that, i.e., even for the initial start of the stem. That way there'd be only one code path that'd always be exercised; with the current scheme, the code path through ZEEK_STEM will be pretty rarely hit (hopefully :) and could bit rod over time 

2. If you can think of a way to btest the restarting logic (i.e, `kill` a node/stem and see that it's restarted correctly), that would be worth adding. (I did some manual testing and it worked well.)

3. Nit: After factoring out Options.{h,cc} seems we could move `usage()` and `parse_cmdline()` in there too?

4. I suggest we declare the script-level API explicitly as unstable for 3.1, so that we can keep tuning it without worrying about the standard deprecation cycle.

jsiwek commented:
> * When starting the stem process: given that we have use exec + ZEEK_STEM anyways when reviving the stem, we could switch to always using that

I'd like to keep the current behavior since always doing the `exec()` introduces race-condition/uncertainty of the binary being changed out from under us now even on the common usage path (whereas now its isolated to rare case of Stem dying).  Bit-rot isn't too concerning since there's now a unit test for Stem revival.

Ultimately, I see the most desirable implementation as one that has no `exec()` usage and doesn't even require the intermediate Stem, but the thing holding us back for now is global state re-organization -- as soon as we have a way to simply reset/initialize global state, we don't need the Stem process or `exec()`, just `fork()`.

> * If you can think of a way to btest the restarting logic (i.e, `kill` a node/stem and see that it's restarted correctly), that would be worth adding.

Added.

> * Nit: After factoring out Options.{h,cc} seems we could move `usage()` and `parse_cmdline()` in there too?

Moved.

> * I suggest we declare the script-level API explicitly as unstable for 3.1, so that we can keep tuning it without worrying about the standard deprecation cycle.

Done.  Or at least that is noted in the script-wide docs, but not per-function.  Could see that being a general convention we start doing though if you think so, like "Added in v3.1.0.  Stable in v4.0.0."

rsmmr commented:
I'm merging this
